gax/src/main/java/com/google/api/gax/retrying/RetryingContext.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright 2018 Google LLC
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met:
+ *
+ *     * Redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *     * Neither the name of Google LLC nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.google.api.gax.retrying;
+
+import com.google.api.core.BetaApi;
+import com.google.api.core.InternalExtensionOnly;
+import com.google.api.gax.rpc.ApiCallContext;
+import com.google.auto.value.AutoValue;
+import javax.annotation.Nullable;
+
+/**
+ * Context for a retryable operation.
+ *
+ * <p>It provides state to individual {@link RetryingFuture}s via the {@link RetryingExecutor}.
+ */
+@BetaApi
+@InternalExtensionOnly
+@AutoValue
+public abstract class RetryingContext {
+  public static Builder newBuilder() {
+    return new AutoValue_RetryingContext.Builder();
+  }
+
+  public static RetryingContext createDefault() {
+    return newBuilder().build();
+  }
+
+  public static RetryingContext fromCallContext(@Nullable ApiCallContext callContext) {
```

Comment:
Won't it make more sense to move this to `ApiCallContext` (and call it something like `createRetryingContext()`)?. `ApiCallContext` is `@InternalExtensionOnly`. This will also allow to make `RetryingContext` an interface instead of an abstract class (for example `ApiCallContext` is an interface).

Reply:
I can move the construction to apicallcontext, but I don’t think it makes sense to turn this into an interface. It’s a simple value class

Reply:
Looking at this a bit closer, I don't think it makes sense to move the construction into ApiCallContext. I would have to duplicate the construction code in all 3 subclasses: `GrpcCallContext`, `HttpJsonCallContext` and `FakeCallContext`

Reply:
The real problem is `import com.google.api.gax.rpc.ApiCallContext;` which makes `gax.retrying` package dependent on rpc package. That is the main reason why we do not want to use ApiCallContext directly. Please move it to `ApiCallContext`, so ApiCallContext knows about retries, not vice versa.

Reply:
done, but I really think that this is a change for the worse

Reply:
@igorbernstein2 I think @vam-google is just looking to not have things conflated. We're both happy to have `@Nonnull` added in gax, but we'd just prefer that to be a separate PR, not incrementally tacked onto other PRs.

Reply:
Wait, I reacted to quickly, wrong topic.

Reply:
This lets keeping retrying logic call-independent.